design/0003-byop.md

```diff
@@ -0,0 +1,127 @@
+
+Flag interactions are defined below:
+| `prometheus.enabled` | `prometheus.url`              | Result                                                                                                                                                                                                                                           |
+|---------|------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| true    | ""               | Install linkerd-prometheus, and configure control-plane components to use it.                                                                                                                                                                    |
+| false   | "prometheus.xyz" | No linkerd-prometheus, control-plane components will be configured to url                                                                                                                                                                        |
+| true    | "prometheus.xyz" | Install linkerd-prometheus, but stil configure control plane to use different url, useful when linkerd-prometheus is configured to remote-write into long term storage solution at url, Thus  powering dashboard and CLI through long term storage. |
+| false   | ""               | No linkerd-prometheus, and no URL to power dashboards. (Should we even allow this?)                                                                                                                                                              |
+
+### Linkerd Check Tooling
+
+When users get their own Prometheus, It would be really great to have
+some tooling that allows users to check if `scrape-configs` are configured correctly and prometheus is able to get metrics.
+
+`--prometheus` flag can be added to linkerd check to run these tests. The following tests are planned:
+
+- Check if the required scrape-configs are present in proemetheus cm.
+( It should be noted that, the prometheus could be present in an different namespace, etc)
+
+- Check if the common metrics are queriable in prometheus, by using the end-point i.e `prometheus.url` from the `linkerd-values` cm.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+- We already have docs for [exporting metrics](https://linkerd.io/2/tasks/exporting-metrics/). We could update those docs, with remote-write, etc.
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- Should we support all prometheus configs in `linkerd-prometheus`?
+
+- How do we keep track of metrics that the proxies expose, to have consistent check tooling?
+
+- Most third party storage solutions have some auth mechanisms, should we allow that? IMO, we shouldn't as they can have a proxy, etc for cases like that.
```

Comment:
This is a great one to call out. I think we should keep it out of scope for v1. As folks complain that we're missing support for their specific solution we can consider updating to handle that use case.

Reply:
yep, makes sense.